### PR TITLE
YouTube's Fast Forward/Rewind behavior

### LINKED ISSUE #1073: [Suggestion] While seeking, show the video screen more clearly
While fast forwarding or rewinding through a video, the screen is dimmed and prev/pause/next icons are shown in the middle of the screen.

Since we want to look at the video being played while doing this, I think it would be better to not dim the screen, and maybe move the prev/pause/next icons and/or make them smaller.

Or maybe make these configurable?

Thanks for making NewPipe!

### LINKED ISSUE #4264: [Feature Request] YouTube's fast forward/rewind behavior
#### Describe the feature you want

It's a replacement of the current fast forward/rewind functionality. For a demonstration watch this video (https://streamable.com/e/qhuca0) or see the images below.
 
image
image
image

#### Is your feature request related to a problem? Please describe it

No.

#### Additional context

When you try the debug ([app-debug-seek-20200915-02.zip](https://github.com/TeamNewPipe/NewPipe/files/5226312/app-debug-seek-20200915-02.zip)) then you should keep this in mind:

The current dev branch has problems with seeking for many videos. You can verify it by seeking the seekbar and check if there is a delay (visually). I recommend these two videos for testing: https://www.youtube.com/watch?v=xTJcASRxbfI and https://www.youtube.com/watch?v=pXRviuL6vMY
They haven't got that great delay and therefore the UI should response better (less jerky) than on most over videos with this build.

#### How will you/everyone benefit from this feature?

I think it feels more natural when
- 3 fast taps results in 20 seconds seek instead of 10 seconds + normal click (showControls will be called)
- 4 fast taps results in 30 seconds instead of 20 seconds

I'm looking forward to your opinions ;)

### PR DESCRIPTION
**Pull request by vkay94**
<p align="center">
<br>
<a href="https://streamable.com/e/b3eeym">Video Preview</a>
</p>

#### What is it?
- [ ] Bugfix (user facing)
- [x] Feature (user facing)
- [ ] Codebase improvement (dev facing)
- [ ] Meta improvement to the project (dev facing)

#### Description of the changes in your PR
This PR adds multi-double-tap seeking with YouTube's progress indicator style and a small change to alpha animation handling (dynamic duration depending on the view's alpha value).

I've decided to open this PR to discuss it here even if it isn't merge-ready instead of in the issue since the code base is basically done when #4587 was merged and I've reduced the added code to the minimum.

#### Features
- more polished design with fluid transitions
- indicator which shows the current progress
- multi double-tap-to-seek design that allows seeking faster in one flow

#### To-do
- Apply specific handling for popup-videos due to its variable sizes (triangle icon size and text size)

#### Fixes the following issue(s)
- closes #4264 
- closes #1073

#### Relies on the following changes
In the moment there's a design "bug" which results in heavy UI lagging during fast double tapping in some fragments. The reason is that with each seeking the history and playlist fragment is reloaded. Since this feature increases the frequency such high reloading is noticeable.

It could be fixed if the lists fragments apply changes to single items instead of reloading the whole list. But that should be done in another PR, probably by changing to the Groupie library.

#### APK testing (2020-12-01)
When you try it, please make sure you haven't opened the history or playlist detail fragments opened (also in tabs).
[Releases page](https://github.com/vkay94/NewPipe/releases/tag/YT-Rewind-Forward)

Here are some visuals for the 2020-12-01-version with its states:

<p align="center">
<br>
<em>Normal video frame for reference</em>
</p>

<p align="center">
<br>
<em>Playing + dark arc shadow</em>
</p>

<p align="center">
<br>
<em>Paused + controls were visible (shadow + dark transparent arc)</em>
</p>

So let's continue discussing it here with the view of the code implementation ;)

#### Due diligence
- [x] I read the [contribution guidelines](https://github.com/TeamNewPipe/NewPipe/blob/HEAD/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by opusforlife2** [comment]
Nice, your PR has a very professional feel to it. :P

**Comment by opusforlife2** [comment]
Yup. Everything is smooth and awesome and absolutely amazeballs, as expected. (I tested the release version).

The only thing that could make it better would be removing that background shadow. :*

**Comment by vkay94** [comment]
@opusforlife2 Preferably I'd like to add a setting to give a opportunity to adjust it individually - why doesn't Github provide a poll functionality xD
I'll go through the issue and assess again (it's a while ago).

Besides of that I've increased the time for the triangle animation cycle a little bit. What do you think?

**Comment by opusforlife2** [comment]
I don't remember how long it took in previous builds. The current one looks fine, though.

> why doesn't Github provide a poll functionality xD

You could say:

- want dark shadow - use hooray emoji on this comment
- want it gone - use eyes emoji on this comment.

or something like that.

**Comment by vkay94** [comment]
@opusforlife2 This option is not really "optimal" :) It's a little bit difficult to decide when there are only a few people voting (3 votes aren't very meaningful - just remember the complaint when the unified player was introduced o.0). The shadow definitely ensures readability under all circumstances.

In the issue back then I made a small "config" button but stopped it due to niche. Now I thought I could implement it for testing - why not ^^:

<p float="left">
  
   
</p>

[app-release-setting-test_20201111.zip](https://github.com/TeamNewPipe/NewPipe/files/5525371/app-release-setting-test_20201111.zip)

Note: The setting is not saved (changes won't be applied to the video player itself).

**Comment by opusforlife2** [comment]
No shadow still seems better to me. Why doesn't the universe like no shadow? ( ;´༎ຶД༎ຶ`)

**Comment by Stypox** [comment]
Oh, if you are talking the dark shadow covering all of the screen (and not only the arc background), then I agree with you @opusforlife2 that the shadow is unneeded. @vkay94 could you rebase? I will then review ;-)

**Comment by vkay94** [comment]
Okay, fine. I'll set it like the following (where _shadow_ is the screen covering shadow, _arc_ the rounded shadow with text and ,_transparent arc_ the one with less transparency shadow):

**State before double tapping => what happens visually when beginning:** (arc is always animating in)
- Video is playing + shadow/controls are hidden => ---
- Video is playing + shadow/controls are visible => animate out shadow/controls
- Video is paused + shadow/controls are hidden => --- **
- Video is paused + shadow/controls are visible => animate out controls (+arc is transparent arc here)

**State during double tapping => what happens visually when it's finished:** (arc is always animating out)
- Video is playing + arc is visible => ---
- Video is paused + arc is visible => animate in shadow + animate in controls
- Video is paused + transparent arc and shadow are visible => animate in controls

**: In the previous versions I animated in the shadow for all cases when the video is paused even if the shadow wasn't visible then. Since the shadow is unwanted that much ( :( ) I'll leave it out in this specific case I guess but I need to check how the animation looks when the arc is animating out while the shadow and controls are animating in simultaneously).

@Stypox Yeah I'll rebase it and commit the above change. There was no need to do it since the debug app works (and I don't like to force-push when it's unnecessary + the feature isn't merge-ready anyway due to the update problem).

**Comment by vkay94** [comment]
@opusforlife2 @Stypox I uploaded a version with the configs above and updated the OP with the testing apk. Let me know whether the dark arc color (second screenshot) is fine ... (for me it's too light for reading, but darker make it unsuitable for other videos xD)

**Comment by opusforlife2** [comment]
YAY NO SHADOW

Bugs found:

- When paused, the seekbar doesn't update after double tapping. It only updates upon tapping Play.

- Pause. Rewind to the beginning. Now double tapping doesn't work at all, either forwards or backwards. It starts working again after you tap Play.

**Comment by Stypox** [comment]
Shapes and colors look good and polished to me now :-D
I found a bug: while playing, tap on the screen to show controls, then double tap on one side of the screen to fast-forward/rewind. The arc appears on the double-tapped side of the screen, as expected, but the `X seconds` text and the triangles animation appears on the opposite side.

**Comment by vkay94** [comment]
>When paused, the seekbar doesn't update after double tapping. It only updates upon tapping Play.

I can reproduce it rarely indeed. I may look into it but I can't find a reason why it happens with the changes made (I call the `onFastForward`/`onFastForward` the same way as before and it calls `triggerProgressUpdate` which updates the progress bar and text). 
The only circumstances when this call does "nothing" (= is not prepared) are when 1. no media found or unavailable or 2. the video is finished/ended. @opusforlife2 Can you do a favor and check whether you can reproduce it in 0.20.3 too?

>Pause. Rewind to the beginning. Now double tapping doesn't work at all, either forwards or backwards. It starts working again after you tap Play.

That was (partially) designed to work like this, so that you can't rewind endlessly. I added a small puffer that you can rewind only when the stream progress is greater or equal of a half second. But I'll add a separate check for the forwarding in this case ;)

>I found a bug: while playing, tap on the screen to show controls, then double tap on one side of the screen to fast-forward/rewind. The arc appears on the double-tapped side of the screen, as expected, but the X seconds text and the triangles animation appear on the opposite side.

Unfortunately, I can't reproduce it, works as expected for me. Could you give me some more information please? (E.g. Android Version or whether it happens for all streams or only rarely). Thank you.

@Stypox @opusforlife2 Thanks for your bug reports :)

**Comment by opusforlife2** [comment]
> Can you do a favor and check whether you can reproduce it in 0.20.3 too?

**Comment by Stypox** [comment]
Here is a video of the bug:
SVID-20201121-142813-1

**Comment by vkay94** [comment]
Hmm, I tried to reproduce it under the same conditions as you show in the video (locked orientation, navigation bar enabled and same video). I tried it on two devices with Android 8.0 and Android 10.0, but it doesn't happen.

I've looked into the code and there might be a race condition between changing the position and animating. @Stypox  Short question since it's hard to see in the gif: When it appears and you keep seeking, does it change the position or does it keep on the same side for this series of taps?

I did a small change which should fix the race condition (I apply the correct position before animating in). Could you test it please? [app-debug-frff-test-change-position.zip](https://github.com/TeamNewPipe/NewPipe/files/5577955/app-debug-frff-test-change-position.zip)

Thanks

**Comment by Stypox** [comment]
@vkay94 if I keep tapping it keeps happening (the side is not correctly set even after skipping multiple tens of seconds), and the apk you provided is no different unfortunately

**Comment by vkay94** [comment]
@Stypox That's weird. I've converted your GIF to a video and analyzed the taps. The arc side is always correct as I see and also the direction of the triangles. The only thing which is incorrect is the position of the view itself. I'm changing it via constraints programmatically. I did some changes to the method, hopefully it helps:
[app-debug-frff-test-change-position-2.zip](https://github.com/TeamNewPipe/NewPipe/files/5579672/app-debug-frff-test-change-position-2.zip)

May I ask you which Android version you're running? @opusforlife2 Could you check whether you encounter Stypox issue as well, please?

**Comment by opusforlife2** [comment]
@vkay94 Already tried it yesterday. I couldn't reproduce it. Could be Huawei specific.

**Comment by vkay94** [comment]
If it's a manufactor-specific issue then I can't do much because I haven't got such device here. But if the above change doesn't fix it, I have one option at least to try :)

**Comment by Stypox** [comment]
@vkay94 with your last apk I can't reproduce it anymore. It was a bug with constraint layout, it seems like

**Comment by vkay94** [comment]
Great to hear that 👍  (that means I've to update my library in the next time too :')).

**Comment by Stypox** [comment]
@vkay94 could you rebase this? I would like to take a look at it soon ;-)

**Comment by vkay94** [comment]
Yeah, I'll take a look at it this weekend 👍

**Comment by vkay94** [comment]
@Stypox I've rebased it (sorry for the delay, I became ill the last days and couldn't do anything)

File: app/src/main/res/layout/player_seek_seconds_view.xml
Comment (Stypox) [inline app/src/main/res/layout/player_seek_seconds_view.xml:2]: Would it be possible to use a `ConstraintLayout` instead, so that there is less view nesting?

File: app/src/main/res/layout/player.xml
Comment (Stypox) [inline app/src/main/res/layout/player.xml:730]:
```suggestion
        android:alpha="0" /> <!-- Required for the first appearance to fade correctly -->
```

**Review by Stypox — CHANGES_REQUESTED** [review]
Thank you! No problem if you were ill ;-)

File: app/src/main/java/org/schabi/newpipe/ktx/View.kt
Comment (vkay94) [reply app/src/main/java/org/schabi/newpipe/ktx/View.kt:59]: You're right. I used this check before moving the check into the view's class. I've copied it during rebasing. Removed it now and it works as it used to.

File: app/src/main/java/org/schabi/newpipe/player/Player.java
Comment (vkay94) [reply app/src/main/java/org/schabi/newpipe/player/Player.java:570]: Sure (but I doubt that there'll be such case in the future ^^)

File: app/src/main/res/layout/player_seek_seconds_view.xml
Comment (vkay94) [reply app/src/main/res/layout/player_seek_seconds_view.xml:2]: Probably I could also try to use `<merge>` since it's a `LinearLayout` (I don't use constraints or anything similar) + extend from `LinearLayout`. I look into it, but I think it should work.

**Comment by opusforlife2** [comment]
Anything to test here or still the same as earlier?

**Comment by vkay94** [comment]
@opusforlife2 The applied changes don't differ from the ones before the rebase, at least for the user experience. 

The fragment's items refresh rate is still the problem (probably I should convert it to draft). 

I already did some work locally regarding it by converting the logic into a more async refresh concept using the Groupie library. But the change is very huge and basically every ListFragment with stream items has to be converted. 

Thankfully, the ViewBinding PRs reduce much of chang s :)

**Comment by Stypox** [comment]
@vkay94 feel free to open a draft pr for that even if you don't have something ready, or even if it doesn't even build. I would like to provide some feedback and maybe some code, as I did something similar to lists for #2583, so we could refactor the whole list thing once and for good.

**Comment by vkay94** [comment]
@Stypox Sure, in the next time but in the moment it's not well structured except of the _core classes_ + the fragments require a rebase. 

If you're interested, I've copied the essential 3 files here in [my gist](https://gist.github.com/vkay94/78f6fb2126cf2b39edda483f316aa3c1) on which the fragments and items are build on.

**Comment by opusforlife2** [comment]
_Whoa_. I've been eagerly waiting for this PR to get updated, but didn't want to nag. This is my daily driver. :D

Can I download the latest APK or will you be making some more changes soon?

Also, make sure you log on to Matrix at least once in 30 days, @vkay94, or the IRC bot kicks you out. You'll have to join the channel again.

**Comment by Atemu** [comment]
Needs a rebase.

**Comment by avently** [comment]
@vkay94 hey, what's the status of the PR? I would like to see it merged ASAP since it's so useful

**Comment by Stypox** [comment]
I rebased, but then when testing on API 19 I could see no arrows

**Comment by Stypox** [comment]
Ok, I've fixed this in the latest commit by using `AppCompatImageView`.

**Comment by opusforlife2** [comment]
Tested the latest APK. Works perfectly fine. 👍

**Comment by tsiflimagas** [comment]
This feature works fine indeed and is one of my favorites! The only thing I could find is that the opaque circle around the play/pause button doesn't show up.
 
(The circle I'm talking about)
Screenshot_20210806-083910048~01

Also, what about that to-do note? Is it still pending? Just wondering if this pr is considered complete.

**Comment by Atemu** [comment]
@Stypox all of your comments are marked as resolved but your review is still blocking merging. I think you have to explicitly approve.

The oddity @tsiflimagas is odd indeed but not anything critical I'd say.  
Popup videos look just fine in my testing too.

Is there anything else that needs to be done before this can be merged?

**Comment by tsiflimagas** [comment]
There's another problem I found with this. If you tap in the player in order for the seekbar, play/pause etc controls to appear, if you switch player (main to popup or the opposite) while they're still showing, the shade (the dark, transparent tint, if that's not the right word) which also appears with the controls, will remain in the newly opened player. Tapping on the player to bring up the controls will make the shade to hide normally along with them.

I really wants this to be merged too!

**Review by Stypox — APPROVED** [review]
@tsiflimagas thank you for testing! I've fixed the issues you found out
- *The only thing I could find is that the opaque circle around the play/pause button doesn't show up.* - fixed in cd45cad1c67c68c2c402f3c76983d497c55e261e (this was pretty tricky)
- *the shade will remain in the newly opened player* - fixed in 8cd30b230a462c4411a10c0974c75b2511b19bed
- @TobiGr Use view binding instead of kotlin synthetics - 4cb7866e1ce3a05fbd73aefed2ff9ca90dd31635

> Popup videos look just fine in my testing too.

I agree. I think this is ready. I've played with @vkay94's code a bit and it does seem to look good :-)

**Comment by tsiflimagas** [comment]
It works great, thank you! 🎉

**Comment by Stypox** [comment]
@TobiGr when you have some time please check this :-)

**Review by TobiGr — APPROVED** [review]

File: app/src/main/java/org/schabi/newpipe/player/Player.java
Comment (litetex) [inline app/src/main/java/org/schabi/newpipe/player/Player.java:89]: Wrong import, checkout ``import androidx.annotation.Nullable;`` some lines above.

**Comment by avently** [comment]
we are closer

**Review by litetex — CHANGES_REQUESTED** [review]
Thank you for the PR!

I just did a quick code review and no testing...

Please add in general a few more comments and javadocs - in particular to classes.

I also don't know if calling it ``PlayerSeekOverlay`` is a good name, as we already have a overlay for the player-seek-bar which shows previews of certain moments in the video.

**Comment by SameenAhnaf** [comment]
When animation settings are turned off, the overlay feels like electric shock. No issues if they were turned on.

https://user-images.githubusercontent.com/79245641/135759727-ce7470b1-060d-4d7c-b3d0-a782fd2134ad.mp4

**Comment by SameenAhnaf** [comment]
#1073 might be linked too.

**Comment by Atemu** [comment]
Rebased here: https://github.com/Atemu/NewPipe/tree/youtube-rewind-forward

**Comment by avently** [comment]
@TobiGr still no merge. What's wrong with this PR?

**Comment by litetex** [comment]
@avently 
> still no merge. What's wrong with this PR?

grafik

**Comment by avently** [comment]
@litetex so what are you waiting for?
https://github.com/TeamNewPipe/NewPipe/commit/7d96c050d6c8f2635293a873c94d00959862c137
Tobi fixed the problem you mentioned.

**Comment by litetex** [comment]
@avently 
https://github.com/TeamNewPipe/NewPipe/pull/4833#pullrequestreview-769509635
> I just did a quick code review and no testing...

I will test it when I have time 😄

**Comment by avently** [comment]
> I will test it when I have time 😄

Anniversary of this PR is coming )

**Comment by opusforlife2** [comment]
> Anniversary of this PR is coming

😆 @vkay94 You wanna wish happy anniversary to your PR like @fynngodau did for his Bandcamp PR?

**Comment by litetex** [comment]
> Anniversary of this PR is coming )

grafik

Next year this issue will be a century old: https://github.com/libgit2/libgit2sharp/issues/229 😆
(RIP to all the PR that are still open)

<sub>Please send more PR reviewers.<br/>In the meantime have fun with this: https://www.google.com/search?q=pull+request+review+memes&tbm=isch</sub>

**Comment by litetex** [comment]
Status update: I'm currently working on this PR.

Checklist:
- [x] Minify / Refactor code (remove old fast-seek UI elements, rename files, make code more understandable)
- [ ] Find a solution so that the user knows where he/she currently is (e.g. at 1:23) - YT shows the seekbar for that
- [ ] Refactor overlay/shadow animations

**Comment by litetex** [comment]
@SameenAhnaf 
Turned off animations should now be respected :)

I also refactored the double tab so that it now works like the current application (shows/hides nothing else when seeking)

File: app/src/main/java/org/schabi/newpipe/views/player/PlayerFastSeekOverlay.kt
Comment (Stypox) [inline app/src/main/java/org/schabi/newpipe/views/player/PlayerFastSeekOverlay.kt:125]: Shouldn't these be `onDoubleTap` instead of `Tab`?

File: app/src/main/java/org/schabi/newpipe/views/player/SecondsView.kt
Comment (Stypox) [inline app/src/main/java/org/schabi/newpipe/views/player/SecondsView.kt:37]: Also here
```suggestion
    // Done as a field so that we don't have to compute on each tap if animations are enabled
```

**Review by Stypox — CHANGES_REQUESTED** [review]

**Comment by litetex** [comment]
So I think this is now ready for a review/test/merging.

The solution with showing the seekbar should better be done in another issue/PR to not bloat up this PR.

Animation also seems to be okay now.

**Comment by opusforlife2** [comment]
> So I think this is now ready for a review/test/merging.

Woohoo! Let's go! I use this regularly and haven't found any issues. I'll test this latest version as well.

**Review by Stypox — CHANGES_REQUESTED** [review]
Great! Good to merge in my opinion. Just this little thing, but after solving it don't wait for my approval and just merge, if there aren't any objections by somebody else

File: app/src/main/java/org/schabi/newpipe/views/player/PlayerFastSeekOverlay.kt
Comment (Atemu) [reply app/src/main/java/org/schabi/newpipe/views/player/PlayerFastSeekOverlay.kt:128]: An enum class would probably be clearer.

Comment (litetex) [reply app/src/main/java/org/schabi/newpipe/views/player/PlayerFastSeekOverlay.kt]: Done

**Review by Stypox — APPROVED** [review]

**Comment by opusforlife2** [comment]
I wanna click on Merge so bad.

**Comment by Stypox** [comment]
@opusforlife2 If you test one last time that everything works well (testing many possible situations), you are allowed to press Merge :-D

**Comment by opusforlife2** [comment]
Oh no. That could take weeks! :O

**Comment by litetex** [comment]
@opusforlife2 @Stypox
⏰

**Comment by litetex** [comment]
Improved one thing:
When the playback has ended (end-screen) it's now possible to rewind.

**Review by Stypox — DISMISSED** [review]
I tested fullscreen, landscape, vertical videos, specific player states (e.g. just after switching player, or soundcloud music videos, etc.) on API 19 emulated, API 31 emulated and API 29 on-device and couldn't find any issue. @litetex if we want to include this in this release, merge!

**Comment by Stypox** [comment]
Wait, I forgot to test changing the settings... and found an issue. When a video is already playing and the fast-forward duration setting is changed, the player still fast-forwards by the old amount until the app is restarted. This does not happen on current `dev`.

**Comment by opusforlife2** [comment]
> and found an issue

**Comment by litetex** [comment]
> Wait, I forgot to test changing the settings... and found an issue. When a video is already playing and the fast-forward duration setting is changed, the player still fast-forwards by the old amount until the app is restarted.



Before it didn't show any amount so I think this works as expected.
Let's simply restart the app when the user changes the duration.
...

Just kidding. I will fix it.

Already found the bad line:
https://github.com/TeamNewPipe/NewPipe/pull/4833/files#diff-68cd2f051cf461f1cee8f97ab4a793ab90276943eed26f8f1baf5105948b2b5eR589

**Review by Stypox — APPROVED** [review]
Works, and the fast-inexact-seek also works ;-)

**Comment by avently** [comment]
Guys, let me know if you need help in clicking on merge button. I have skills

**Comment by avently** [comment]
Thank you to all contributors!

**Comment by opusforlife2** [comment]
Is this real or am I dreaming?

**Pull request merged**
